xapian-applications/omega/worker.cc

```diff
@@ -0,0 +1,185 @@
+/** @file worker.cc
+ * @brief Class representing worker process.
+ */
+/* Copyright (C) 2005,2006,2007,2008,2009,2010,2011 Olly Betts
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "worker.h"
+#include "worker_comms.h"
+
+#include "pkglibbindir.h"
+#include <csignal>
+#include <cstring>
+#include <cerrno>
+#include "safefcntl.h"
+#include "safeunistd.h"
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include "closefrom.h"
+#include "freemem.h"
+
+using namespace std;
+
+bool Worker::ignoring_sigpipe = false;
+
+void
+Worker::start_worker_subprocess()
+{
+    int fds[2];
+    if (socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC, fds) < 0) {
+	throw string("socketpair failed: ") + strerror(errno);
+    }
+
+    child = fork();
+    if (child == 0) {
+
+	// Child process.
+	close(fds[0]);
+
+	// Connect pipe to fd 3.  Don't use stdin and stdout in case the
+	// filter library tries to read from stdin or write debug or progress
+	// info to stdout.
+	dup2(fds[1], 3);
+
+	// Make sure we don't hang on to open files which may get deleted but
+	// not have their disk space released until we exit.
+	closefrom(4);
+
+	// Connect stdin, stdout, stderr to /dev/null.
+	int devnull = open("/dev/null", O_RDWR);
+	dup2(devnull, 0);
+	dup2(devnull, 1);
+	dup2(devnull, 2);
+	if (devnull > 3) close(devnull);
+
+	// FIXME: For filters which support a file descriptor as input, we
+	// could open the file here, and pass the file descriptor across the
+	// socket to the worker process using sendmsg().  Then the worker
+	// process could be chroot()-ed to a sandbox directory, which means
+	// we're reasonably protected from security bugs in the filter.
+
+	const char * mod;
+	if (filter_module.find('/') == string::npos && false) {
```

Comment:
`&& false`?

Reply:
If that was for testing in an uninstalled tree, you can probably achieve what you want using `export XAPIAN_OMEGA_PKGLIBBINDIR=$PWD` in the shell you're testing from - this overrides what `get_pkglibbindir()` returns.

Reply:
Sorry, I forget about that. I will try to be more careful.